### DISCUSSION TITLE
Use ZodOptional in input

### DISCUSSION THREAD

**Discussion by JonathanCabezas** [RobinTail/express-zod-api#1992] [Q&A] [answered]
Hello,

I'm trying to use `optional` to avoid code redundancy with zod, but it doesn't seem to pass type-checking in endpoint inputs.

Here my code: 
```

export const userIDschema = z.object({
  user_id: idSchema,
});

export const userSchema = userIDschema.extend({
  first_name: z.string(min1),
  last_name: z.string(min1),
  mail: z.string().email(),
  date_registration: z.string().datetime()
});

export const updateUserEndpoint = routeFactory
  .addMiddleware(authMiddleware(Permission.Write, Modules.User))
  .build({
    method: "patch",
    tag: "user",
    description: "Change an user information by id.",
    input: userSchema.pick({ first_name: true, last_name: true, mail: true }).optional().extend({
      user_id: greaterThanZeroIntFromStringSchema,
    }),
    output: emptySchema,
    handler: async ({ input }) => {
      if (!(await updateUser(input))) {
        throw userNotFoundError;
      }

      return {};
    },
  });
```

Did I miss something ?

**Comment by RobinTail** [comment, marked as answer]
You are making the whole object schema optional.
Consider making optional the fields inside of it instead, @JonathanCabezas

**Reply by RobinTail** [reply]
Try `.partial()` instead of `.optional()`, @JonathanCabezas 

`input` is an aggregation of `request` properties — it's always an object, therefore it can not pass `.optional()` because it implies `undefined` which `input` can not be.

![](https://raw.githubusercontent.com/RobinTail/express-zod-api/master/dataflow.svg)

**Reply by JonathanCabezas** [reply]
I feel ashamed because I meant to use `partial` but I confused myself.

Thanks for the help, it works as expected with right function.
